analyzer/src/traversal/module.rs

```diff
@@ -17,6 +19,9 @@ use std::rc::Rc;
 pub fn module(context: Shared<Context>, module: &fe::Module) -> Result<(), SemanticError> {
     let scope = ModuleScope::new();
 
+    // Walk over the module but ignore everything that is *inside* contracts.
+    // This is to make contract types known prior to where we know their exact shape
+    // so that they can already be used as types in other contracts.
     for stmt in module.body.iter() {
```

Comment:
To reiterate, I think there should only be two passes here: a `def` pass and a `body` pass. The `def` pass would be responsible for adding definitions to the scope and the body `pass` would be responsible for analyzing function bodies (in both structs and contracts). To analyze definitions, we shouldn't need to dive into any bodies, but for function bodies, we will need information from the definition pass.